src/thread.rs

```diff
@@ -210,39 +196,49 @@ pub struct ScopedJoinHandle<'scope, T: 'scope> {
 unsafe impl<'scope, T> Send for ScopedJoinHandle<'scope, T> {}
 unsafe impl<'scope, T> Sync for ScopedJoinHandle<'scope, T> {}
 
-/// Create a new `Scope` for [*scoped thread spawning*](struct.Scope.html#method.spawn).
+/// Creates a new `Scope` for [*scoped thread spawning*](struct.Scope.html#method.spawn).
 ///
-/// In addition, you can [register ad-hoc functions](struct.Scope.html#method.defer) that are
-/// deferred to be run. No matter what happens, before the `Scope` is dropped, it is guaranteed that
-/// all the unjoined spawned scoped threads are joined and the deferred functions are run.
+/// No matter what happens, before the `Scope` is dropped, it is guaranteed that all the unjoined
+/// spawned scoped threads are joined.
 ///
-/// `thread::scope()` returns `Ok(())` if all the unjoined spawned threads and the deferred
-/// functions did not panic. It returns `Err(e)` if one of them panics with `e`. If many of them
-/// panics, it is still guaranteed that all the threads are joined and all the functions are run,
-/// and `thread::scope()` returns `Err(e)` with `e` from a panicking thread or function.
+/// `thread::scope()` returns `Ok(())` if all the unjoined spawned threads did not panic. It returns
+/// `Err(e)` if one of them panics with `e`. If many of them panic, it is still guaranteed that all
+/// the threads are joined, and `thread::scope()` returns `Err(e)` with `e` from a panicking thread.
 ///
 /// # Examples
 ///
 /// Creating and using a scope:
 ///
 /// ```
 /// crossbeam_utils::thread::scope(|scope| {
-///     scope.defer(|| println!("Exiting scope"));
+///     scope.spawn(|| println!("Exiting scope"));
 ///     scope.spawn(|| println!("Running child thread in scope"));
 /// }).unwrap();
-/// // Prints messages
 /// ```
 pub fn scope<'env, F, R>(f: F) -> thread::Result<R>
 where
     F: FnOnce(&Scope<'env>) -> R,
 {
     let mut scope = Scope {
-        dtors: RefCell::new(None),
+        joins: RefCell::new(Vec::new()),
+        panics: RefCell::new(Vec::new()),
         _marker: PhantomData,
     };
-    let ret = f(&scope);
-    scope.drop_all()?;
-    Ok(ret)
+
+    // Executes the scoped function. Panic will be catched as `Err`.
+    let result = panic::catch_unwind(panic::AssertUnwindSafe(|| f(&scope)));
+
+    // Joins all the threads.
+    scope.drop_all();
+    let panic = scope.panics.borrow_mut().pop();
```

Comment:
Before this line, we could do:

```rust
let result = match result {
    Ok(r) => r,
    Err(e) => panic::resume_unwind(e),
};
```

And return the result as `Ok(result)` at the end of function.

This way we only catch panics inside spawned threads, not inside the scope itself.

Do you agree with this, or do you think we should catch panics inside the scope, too? I'm personally not 100% sure yet...

Reply:
Yeah, following whatever Rayon does makes sense.

I'll also ask one more person for opinion, `prz` on IRC, who's been following recent developments in `crossbeam-utils`.

Reply:
So there are a number of differences between scopes in crossbeam and rayon:

* In crossbeam, spawned threads can return a result. In rayon, they can't.
* In crossbeam, spawned threads can be manually joined. In rayon, they can't.
* In crossbeam, panics from spawned threads can be handles using `.join()`. In rayon, they can't.
* In crossbeam, panics from unjoined threads can be handled as the `Result` returned from `scope`. In rayon, `scope` simply propagates all panics and returns a `T`, not `Result`.

Basically, it looks like in rayon all panics are propagated, while in crossbeam panics tend to be caught and handled manually.

Reply:
I agree -- let's return `Result` instead of propagating panic.  I'm merging it.  Thanks!